File: 239-sliding-window-maximum/sliding-window-maximum.cpp
#include<deque>
#include<vector>

class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
       deque<int> q;
       vector<int> result;
       int l = 0;
       int r = 0;

       while(r < nums.size()){
            while (q.size() && nums[q.back()] < nums[r] ){
                q.pop_back();
            }
            q.push_back(r);

            if (l > q.front()){
                q.pop_front();
            }

            if (r + 1 >= k){
                result.push_back(nums[q.front()]);
                l += 1;
            }
            r += 1;
       }

       return result; 
    }
};

File: 239-sliding-window-maximum/sliding-window-maximum.java
import java.util.Deque;
import java.util.ArrayDeque;

class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        Deque<Integer> q = new ArrayDeque<Integer>();
        int result[] = new int[nums.length+1-k];
        int l = 0;
        int r = 0;
        while(r < nums.length){
            while ( q.size() > 0 && nums[q.peekLast()] < nums[r]) q.removeLast();
            q.addLast(r);

            if (l > q.peekFirst()) q.removeFirst();

            if (r+1 >= k){
                l += 1;
                result[(r+1) - k] = nums[q.peekFirst()];
            }
            r += 1;
        }
        return result;
    }
}